Cleanup API spec

matthewhanson opened the pull request:
Initial draft, still need to clean up and expand the documentation in README.

This PR:

- consolidates documentation into README.md  and search.md. There was a bunch of redundant or outdated information
- Improve process and documentation on creating OpenAPI from existing definitions, fragments, extensions

cholmes commented:
Hey @matthewhanson - just realized we have an approved PR for some API stuff - https://github.com/radiantearth/stac-spec/pull/415 I was about to merge, but then thought I should pause in case it might mess up this PR.

m-mohr commented on api-spec/README.md:
```diff
@@ -1,100 +1,51 @@
+In the definitions directory there are two files: WFS3.yaml and STAC.yaml. The WFS3.yaml file is the WFS3 OpenAPI definition **as currently used by STAC**. The STAC.yaml file contains additional endpoints and components that STAC uses, which is treated as a WFS3 extension. A basic STAC implementation would implement both the WFS3 and STAC specifications.
 
-See [yaml-files](https://www.npmjs.com/package/yaml-files) for the syntax to import YAML fragments.
+The YAML files in the extensions dolder are fragments. Fragments are used to describe incomplete pieces of an OpenAPI document, and must be merged with a complete OpenAPI document to be usable. This way extensions can be kept separate, and implementors can combine just the extensions they want to use in order to create a custom OpenAPI document that can be used to generate docuemntation.
```
dolder => folder

matthewhanson replied:
fixed

matthewhanson commented:
This is a big PR, so it might be time consuming to properly review. Since it changes pretty much the entire api-spec directory it's probably easiest to just view that branch and look through the api-spec folder. A few changes were made outside that directory, mostly just to remove redundant or outdated information about the API.

Notable changes:

- Redone README, consolidated info various files
- examples added
- API spec info separated into api-spec.md
- Process of creating OpenAPI docs clarified, given in README
- Refactored OpenAPI docs and fragments
- Moved transaction extension into API extensions
- API extensions now each have their own folder and README, with an extensions README similar to the content extensions README

Resolves #297 
Resolves #203

m-mohr commented on api-spec/extensions/README.md:
```diff
@@ -0,0 +1,132 @@
+
+## List of community extensions
+
+| Extension Name (Prefix)                                      | Description                                                  | Maturity |
```
I think we can remove the "(Prefix)".

m-mohr commented on api-spec/extensions/transaction/README.md:
```diff
@@ -1,6 +1,12 @@
-**Extension [Maturity Classification](../README.md#extension-maturity): Proposal**
+**Extension [Maturity Classification](../../../extensions/README.md#extension-maturity): Pilot**
+
+This document explains the fields of the STAC Electro-Optical (EO) Extension to a STAC Item. EO
```
Copy&Paste? ;-)

m-mohr commented on api-spec/api-spec.md:
````diff
@@ -0,0 +1,47 @@
+STAC provides some additional endpoints for the root Catalog itself, as well as the capability to search the Catalog. Note that a STAC API does not need to implement WFS3, in which case it would only support the endpoints given below. See the [OpenAPI specification document](definitions/STAC-standalone.yaml).
+
+```
+| Endpoint      | Returns          | Description        |
````
Was this meant to be a table? It is rendered as code...

m-mohr commented:
I added some comments and have some more thing I wasn't able to add review comments for:

* The link to the implementation.md in api-spec/extensions/transaction/README.md is broken.
* Add a prominent link from the main extension page to the API extensions.

m-mohr commented on api-spec/api-spec.md:
```diff
@@ -0,0 +1,47 @@
+
+| Parameter      | Type          | Description        |
+| ------------ | ------------- | ---------------------- |
+| bbox | [number] (x4)        | Requested bounding box [west, south, east, north] |
```
What does the (x4) mean? What about the z axis?

cholmes replied:
+1

m-mohr commented on api-spec/api-spec.md:
```diff
@@ -0,0 +1,47 @@
+| bbox | [number] (x4)        | Requested bounding box [west, south, east, north] |
+| time | datetime string | Single date, date+time, or a range ('/' seperator) |
+| intersects | GeoJSON Feature | Searches items by performing intersection between their geometry and provided GeoJSON Feature |
+| page | number | The page number of results. Defaults to 1) |
```
The last ) seams to be a left-over...

m-mohr commented on api-spec/api-spec.md:
```diff
@@ -0,0 +1,47 @@
+| Parameter      | Type          | Description        |
+| ------------ | ------------- | ---------------------- |
+| bbox | [number] (x4)        | Requested bounding box [west, south, east, north] |
+| time | datetime string | Single date, date+time, or a range ('/' seperator) |
```
In the rest of the API we just use string as data type and define the string type (datetime) with links to the corresponding RFC in the description. Would just copy that over from the other specs.

cholmes replied:
+1

cholmes commented on api-spec/README.md:
```diff
@@ -1,100 +1,55 @@
+**Examples:** For samples of how the STAC API can be queried, see the *[examples.md](examples.md)* file.
 
-## API Fragments
+**Schemas:** The API is described by the OpenAPI specification documents in the *[openapi](openapi/)* folder.
```
I don't think we need to call this 'schemas', as they aren't really schemas in the sense of JSON or XML schemas. I think we can just say 'OpenAPI documents' or 'OpenAPI fragments', etc.

matthewhanson replied:
From https://swagger.io/docs/specification/about/ it seems like maybe the best term is 'OpenAPI description'.

cholmes commented on api-spec/api-spec.md:
````diff
@@ -0,0 +1,47 @@
+| /stac/search | Items | GeoJSON FeatureCollection of Items found |
+```
+
+The `stac/search` endpoint is similar to the `items` endpoint in WFS3 in that it accepts parameters for filtering, however it performs the filtering across all collections. The parameters accepted are the same as the Filter Parameters above, however the *[extensions](extensions/README.md)* also provide advanced querying parameters.
````
So I'd like to add a line here like: 'The `stac` endpoint should function as a complete `Catalog` representation of all the data contained in the STAC API - all `Item`s should be linked in some way from that root through catalogs/collections, and as `item` links from those catalogs. But then I realized that sat-api isn't doing that yet. 

I feel like we should have an endpoint that stac browser can point at and navigate through down to the items in the same way you can with a static catalog. Though perhaps it makes sense to call that endpoint /browse or something, and then it'd be easier to not make it required?

matthewhanson replied:
STAC Browser can now browse APIs because it can @mojodna added the ability to resolve links with rel=items as well as rel=item. 

I actually removed the existing paragraph about the API sub-catalogs and opened an issue for it:
https://github.com/radiantearth/stac-spec/issues/438

So in short, yes we should discuss how a browsable API should work and can add a better description of it, but we can make it another PR.

matthewhanson commented:
> I added some comments and have some more thing I wasn't able to add review comments for:
> * Add a prominent link from the main extension page to the API extensions.

@m-mohr Not sure what you mean here. The table of API extensions in extensions/README.md does link to each one. Or do you mean something else?

matthewhanson commented:
@m-mohr @cholmes 
Ok I resolved all these issues except for the Browsable API, which is a new issue #438 and a link from the 
main extension page to API extensions.

matthewhanson commented:
Oh, and also didn't update Open API schema to something else yet, but will update that everywhere when we agree what it should be called. 
Rather than "openapi-schema" should it be "openapi-description"?  OpenAPI Schema => OpenAPI Description ?

m-mohr commented:
> 
> 
> > I added some comments and have some more thing I wasn't able to add review comments for:
> > 
> > * Add a prominent link from the main extension page to the API extensions.
> 
> @m-mohr Not sure what you mean here. The table of API extensions in extensions/README.md does link to each one. Or do you mean something else?

I meant the extension page with the content extensions (/extensions/README.md) should link to the API extensions (/api-spec/extensions/README.md).

matthewhanson commented:
Ok  thanks, I added a couple sentences in the extensions/ folder differentiating between content and API extensions and added link.

Also removed "schema" from OpenAPI references.

That takes care of all the comments so far, other than the Browsable API, which is handles by #438
